packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -93,89 +150,78 @@ export const makeZoeHelpers = (zcf) => {
     rejectOffer,
 
     /**
-     * Compare two proposals for compatibility. This returns true
-     * if the left offer would accept whatever the right offer is offering,
-     * and vice versa.
-     *
-     * @param {OfferHandle} leftOfferHandle
-     * @param {OfferHandle} rightOfferHandle
-     * @returns boolean
-     *
+     * Check whether an update to currentAllocation
+     * satisfies offer safety. AmountKeywordRecord is merged with currentAllocation,
+     * overwriting if the keywords are the same to produce the
+     * newAllocation.
+     * @param {OfferHandle} offerHandle
+     * @param {AmountKeywordRecord} amountKeywordRecord
+     * @returns {boolean}
      */
-    canTradeWith: (leftOfferHandle, rightOfferHandle) => {
-      const { proposal: left } = zcf.getOffer(leftOfferHandle);
-      const { proposal: right } = zcf.getOffer(rightOfferHandle);
-      const leftAllocation = zcf.getCurrentAllocation(leftOfferHandle);
-      const rightAllocation = zcf.getCurrentAllocation(rightOfferHandle);
-      const satisfied = (want, availableForTrade) =>
-        getKeywords(want).every(keyword => {
-          const amountMath = zcf.getAmountMath(want[keyword].brand);
-          return amountMath.isGTE(availableForTrade[keyword], want[keyword]);
-        });
-      return (
-        satisfied(left.want, rightAllocation) &&
-        satisfied(right.want, leftAllocation)
-      );
+    isOfferSafe: (offerHandle, amountKeywordRecord) => {
+      const currentAllocation = zcf.getCurrentAllocation(offerHandle);
+      const newAllocation = {
+        ...currentAllocation,
+        ...amountKeywordRecord,
+      };
+      const { proposal } = zcf.getOffer(offerHandle);
+      return isOfferSafeForOffer(zcf.getAmountMath, proposal, newAllocation);
     },
 
     /**
-     * Compare two proposals for compatibility, mapping keywords. This returns
-     * true if the left offer would accept what the right offer is offering,
-     * presuming that the goods at each keyword are transferred across.
+     * Trade between left and right so that left and right end up with
+     * the declared gains.
+     * @param {offerHandleGainsLossesRecord} keepLeft
+     * @param {offerHandleGainsLossesRecord} tryRight
+     * @returns {undefined | Error}
      *
-     * @param {OfferHandle} leftOfferHandle
-     * @param {OfferHandle} rightOfferHandle
-     * @param {[Keyword]} keywords
-     * @returns boolean
+     * @typedef {object} offerHandleGainsLossesRecord
+     * @property {OfferHandle} offerHandle
+     * @property {AmountKeywordRecord} gains - what the offer will
+     * gain as a result of this trade
+     * @property {AmountKeywordRecord?} losses - what the offer will
+     * lose as a result of this trade
```

Comment:
I see now that this does say that `losses` are optional, but the consequences weren't obvious. Would you mind adding something here to explain how this works, and simplifies the case when the keywords match?

Reply:
Done